src/cryptonote_core/blockchain.cpp

```diff
@@ -1823,43 +1886,76 @@ bool Blockchain::create_block_template_internal(
                 expected_reward,
                 b.major_version,
                 height,
-                std::move(l2_range))) {
+                l2_mempool_max))
         return false;
-    }
-    pool_cookie = m_tx_pool.cookie();
+
+    pool_cookie = tx_pool.cookie();
 
     /*
      two-phase miner transaction generation: we don't know exact block weight until we prepare
      block, but we don't know reward until we know block weight, so first miner transaction
      generated with fake amount of money, and with phase we know think we know expected block weight
      */
     // make blocks coin-base tx looks close to real coinbase tx to get truthful blob weight
-    auto miner_tx_context =
-            info.is_miner
-                    ? oxen_miner_tx_context::miner_block(
-                              m_nettype, info.miner_address, m_service_node_list.get_block_leader())
-                    : oxen_miner_tx_context::pulse_block(
-                              m_nettype,
-                              info.service_node_payout,
-                              m_service_node_list.get_block_leader());
+    auto miner_tx_context = info.is_miner ? oxen_miner_tx_context::miner_block(
+                                                    m_nettype,
+                                                    info.miner_address,
+                                                    service_node_list.get_next_block_leader())
+                                          : oxen_miner_tx_context::pulse_block(
+                                                    m_nettype,
+                                                    info.service_node_payout,
+                                                    service_node_list.get_next_block_leader());
+
+    if (hf_version >= cryptonote::feature::ETH_TRANSITION) {
+        std::memcpy(
+                b.sn_winner_tail.data(),
+                miner_tx_context.block_leader.key.data() +
+                        miner_tx_context.block_leader.key.size() - b.sn_winner_tail.size(),
+                b.sn_winner_tail.size());
+    }
+
+    if (hf_version >= cryptonote::feature::ETH_BLS) {
+        auto actual_reward = m_l2_tracker->get_reward_rate(b.l2_height);
+        if (!actual_reward) {
+            log::error(
+                    logcat,
+                    "L2 Tracker failed to retrieve the current block reward; unable to create a "
+                    "block");
+            return false;
+        }
+
+        // Clamp the l2_reward value into the maximum allowed range by the chain (this means sudden
+        // pool reward changes will only show up gradually in l2_reward, but prevents abuse by
+        // faulty/compromised L2 providers or pulse quorums).
+        auto l2r_range = l2_reward_range(height);
+        b.l2_reward = std::clamp(*actual_reward, l2r_range.first, l2r_range.second);
```

Comment:
We aren't setting `b.rewards` here and I couldn't see any other place where it is being set for HF21 blocks.
```cpp
        b.reward  = eth_consensus_reward(height);
```
I haven't figured out why this isn't triggering the check in `blockchain::validate_miner_transaction` namely this block of code
```cpp
        if (auto expected_reward = eth_consensus_reward(height); b.reward != expected_reward) {
            log::error(
                    log::Cat("verify"),
                    "block reward for height {} is incorrect; block has {} but expected {}",
                    height,
                    print_money(b.reward),
                    print_money(expected_reward));
            return false;
        }
```

Reply:
It didn't get caught because I broke the validation to fix another issue. [28dbb9b](https://github.com/oxen-io/oxen-core/pull/1697/commits/28dbb9b09f214ddc2fa8eb47182b9068c36e92a4) fixes both issues.